File: leetcode/C++/Numberof1Bits.cpp

// Write a function that takes an unsigned integer and returns the number of ’1' bits it has (also known as the Hamming weight).

// For example, the 32-bit integer ’11' has binary representation 00000000000000000000000000001011, so the function should return 3.

class Solution {
public:
    int hammingWeight(uint32_t n) {
        int cnt=0;
        while(n!=0)
        {
            cnt+=n&1; //not cnt=cnt+n&1  because + is prior than &, and & than +=
            n=n>>1;
        }
        return cnt;
    }
};

class Solution{
public:
	int hammingWeight(uint32_t n){
		int cnt=0;
		while(n!=0)
		{
			if(n%2==1) //The end is 1.
				cnt++;
			n/=2;
		}
		return cnt;
	}
};

File: OfferTest/leetcode/src/leetcode/Numberof1Bits.java
package leetcode;

public class Numberof1Bits {

	public static int hammingWeight(int n) {
		
		int cnt=0;
		
		while(n>0){
			cnt++;
			n=(n-1) &n;
		}
		
		return cnt;
    }
	
	public static int hammingWeight2(int n) {
		return Integer.bitCount(n);
    }
	
	
	public static void main(String[] args) {
		
		System.out.println(hammingWeight2(120));
		System.out.println(hammingWeight(120));	
	}

}
